_posts/2019-10-23-kafka-authentication-using-oauth-2.0.md

```diff
@@ -0,0 +1,519 @@
+---
+layout: post
+title:  "Kafka authentication using OAuth 2.0"
+date: 2019-10-2X
+author: marko_strukelj
+---
+
+In Strimzi 0.14 we have added an additional authentication option to the standard set supported by Kafka brokers.
+Your Kafka clients can now use OAuth 2.0 token-based authentication when establishing a session to Kafka broker.
+With this kind of authentication Kafka clients and brokers talk to a central OAuth 2.0 compliant authorization server. 
+Clients use the authorization server to obtain access tokens, or are configured with access tokens issued by the server.
+Brokers communicate with authorization server to validate the tokens presented by clients, thus confirming their identities.
+In this model user creation does not happen on Kafka brokers any more.
+Rather, users are typically managed through identity providers exposed over LDAP.
+Strimzi client components like Kafka Bridge, Kafka Connect, and Kafka Mirror Maker can be configured with OAuth 2.0 authentication through custom resource definitions.
+
+
+<!--more-->
+
+## Authentication in Kafka
+
+When using Apache Kafka you have several options to implement authentication. 
+You can use TLS with client certificates, and both establish a secure channel, and authenticate a client at the same time.
+Another option is to separately configure TLS at connection level, and use a specific SASL protocol for authentication.
+
+There are several SASL based protocols you can use.
+One option is to use SASL PLAINTEXT to authenticate with username and password.
+Another option is to use SASL SCRAM which is also username and password based, but rather than storing clear text passwords it stores salted hashes.
+These two implementations both store usernames and credentials in ZooKeeper.
+
+A more advanced, enterprise-ready solution is to use SASL GSSAPI mechanism which provides support for Kerberos. 
+That typically uses a central LDAP server for managing users, and permissions.
+The problem is that it is quite complex to setup, and people often find it hard to configure.
+
+Since Kafka version 2.0.0 there is an extensible OAuth 2.0 compatible token-based mechanism available, called SASL OAUTHBEARER.
+
+At Strimzi project we have developed extensions, that provide integration with OAuth 2.0 compliant authorization servers. 
+That means that in principle you can use any OAuth 2.0 compliant authorization server to expose centrally managed users for authentication with Kafka.
```

Comment:
It's there because you have to test with each new server to make sure that it really works. When testing with Hydra I found elements of how Hydra creates access tokens that doesn't work with the version of kafka-oauth that made it in 0.14. I had to add some additional config options to accommodate Hydra. It may be the same with other servers. Because we haven't tested, we can say for sure that it works.

Reply:
Thanks for the clarification.